tests/nn/beam_search_test.py

```diff
@@ -232,3 +231,102 @@ def test_empty_sequences(self):
+        ).search(initial_predictions, {}, take_step)
+
+        beam_size = beam_size or 1
+        batch_size = 5
+
+        # top_p should be shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(top_k.size())[:-1] == [batch_size, beam_size]
+
+        assert ((0 <= top_k) & (top_k <= 5)).all()
+
+        # log_probs should be shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(log_probs.size()) == [batch_size, beam_size]
+
+    def test_empty_p(self):
+        initial_predictions = torch.LongTensor([self.end_index - 1, self.end_index - 1])
+        take_step = take_step_with_timestep
+
+        with pytest.warns(RuntimeWarning, match="Empty sequences predicted"):
+            predictions, log_probs = BeamSearch.top_p_sampling(self.end_index, beam_size=1).search(
+                initial_predictions, {}, take_step
+            )
+        # predictions hould have shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(predictions.size()) == [2, 1, 1]
+        # log probs hould have shape `(batch_size, beam_size)`.
+        assert list(log_probs.size()) == [2, 1]
+        assert (predictions == self.end_index).all()
+        assert (log_probs == 0).all()
+
+    def test_empty_k(self):
+        initial_predictions = torch.LongTensor([self.end_index - 1, self.end_index - 1])
+        take_step = take_step_with_timestep
+
+        with pytest.warns(RuntimeWarning, match="Empty sequences predicted"):
+            predictions, log_probs = BeamSearch.top_k_sampling(self.end_index, beam_size=1).search(
+                initial_predictions, {}, take_step
+            )
+        # predictions hould have shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(predictions.size()) == [2, 1, 1]
+        # log probs hould have shape `(batch_size, beam_size)`.
+        assert list(log_probs.size()) == [2, 1]
+        assert (predictions == self.end_index).all()
+        assert (log_probs == 0).all()
+
+    @pytest.mark.parametrize(
+        "k",
+        [-1.0, 1.2, 1.1, "foo", float("inf")],
+    )
+    def test_k_val(self, k):
+        with pytest.raises(ConfigurationError):
+            initial_predictions = torch.tensor([0] * 5)
+            take_step = take_step_with_timestep
+            beam_size = 3
+            top_k, log_probs = BeamSearch.top_k_sampling(
+                self.end_index, k=k, beam_size=beam_size
+            ).search(initial_predictions, {}, take_step)
+
+    @pytest.mark.parametrize(
+        "p",
+        [-1.0, 1.1, 2, "foo", float("inf")],
+    )
+    def test_p_val(self, p):
+        with pytest.raises(ConfigurationError):
+            initial_predictions = torch.tensor([0] * 5)
+            take_step = take_step_with_timestep
+            beam_size = 3
+            top_p, log_probs = BeamSearch.top_p_sampling(
+                self.end_index, p=p, beam_size=beam_size
+            ).search(initial_predictions, {}, take_step)
```

Comment:
Given that `BeamSearch` is now registrable, we should also test instantiating `BeamSearch` `from_params`. A simple test could look like this:

```python
from allennlp.common.params import Params

beam_search = BeamSearch.from_params(Params({"beam_size": 2}))
assert beam_search.beam_size == 2
```

And then also make sure `top_p` and `top_k` can be instantiated correctly `from_params()`.